designs/sam-config.md

```diff
@@ -0,0 +1,369 @@
+```
+
+Why samconfig under `.aws-sam`
+---------------------------------
+
+The `.aws-sam` directory within the project directory is created with normal 755 permissions as default without any special permisions. `sam build` only creates a build directory within `.aws-sam` as `.aws-sam/build`. This directory is erased and re-built on every build. but top level directory is left unaffected.
+
+The `.gitignore` specified in the init apps also only have `.aws-sam/build` ignored and not anything else.
+
+
+Overrides
+----------
+
+The default location of a .aws-sam/samconfig can be replaced by overriding an environment variable called `SAM_CLI_CONFIG`
+
+`
+export SAM_CLI_CONFIG=~/Users/username/mysamconfig.toml
+`
+
+Users can pass an environment `--env` for the section that will be scanned within the configuration file to pass parameters through.
+
+By default the `default` section of the configuration is chosen.
+
+```
+[default]
+
+[default.build]
+[default.build.parameters]
+profile="srirammv"
+debug=true
+skip_pull_image=true
+use_container=true
+
+[default.package]
+[default.package.parameters]
+profile="srirammv"
+region="us-east-1"
+s3_bucket="sam-bucket"
+output_template_file="packaged.yaml"
+
+[default.deploy]
+[default.deploy.parameters]
+stack_name="using_config_file"
+capabilities="CAPABILITY_IAM"
+region="us-east-1"
+profile="srirammv"
+
+```
+
+If a custom environment is specified, the environment is looked up in `samconfig.toml` file instead.
+
+`sam build --env dev`
+
+Sample configuration file
+
+```
+[default.build.paramaters]
+profile="srirammv"
+debug=true
+skip_pull_image=true
+use_container=true
+
+[default.package.parameters]
+profile="srirammv"
+region="us-east-1"
+s3_bucket="sam-bucket"
+output_template_file="packaged.yaml"
+
+[default.deploy.parameters]
+stack_name="using_config_file"
+capabilities="CAPABILITY_IAM"
+region="us-east-1"
+profile="srirammv"
+
+
+[dev.build.paramaters]
+profile="srirammv"
+debug=true
+skip_pull_image=true
+use_container=true
+
+[dev.package.parameters]
+profile="srirammv"
+region="us-east-1"
+s3_bucket="sam-bucket"
+output_template_file="packaged.yaml"
+
+[dev.deploy.parameters]
+stack_name="using_config_file"
+capabilities="CAPABILITY_IAM"
+region="us-east-1"
+profile="srirammv"
+```
+
+
+The configuration file can then be potentially intialized
+
+* all sam init projects could come with a sample samconfig file
+
+
+Config file in Git Repos
+------------------------
+
+`samconfig.toml` file can be checked into a git repo, so that its ready to use on cloning the repo. if the configuration file does not present all the necesssary parameters, the command fails just as if one had specified the same arguments on the command line directly.
+
+Optionally, if multiple configuration files are checked in. One can change the `SAM_CLI_CONFIG` environment variable to point a different configuration file.
+
+`--env` can also be passed in to deal with custom environments defined in the configuration file.
```

Comment:
Could we use a different parameter name than this, something more descriptive/connected to the config file?

Reply:
I'm open to suggestions.